pkg/volume/flexvolume/probe.go

```diff
@@ -186,22 +181,23 @@ func (prober *flexVolumeProber) updateProbeNeeded() {
 func (prober *flexVolumeProber) addWatchRecursive(filename string) error {
 	addWatch := func(path string, info os.FileInfo, err error) error {
 		if info.IsDir() {
-			if err := prober.watcher.Add(path); err != nil {
+			if err := prober.watcher.AddWatch(path); err != nil {
 				glog.Errorf("Error recursively adding watch: %v", err)
 			}
 		}
 		return nil
 	}
-
-	return filepath.Walk(filename, addWatch)
+	return prober.fs.Walk(filename, addWatch)
 }
 
 // Creates a new filesystem watcher and adds watches for the plugin directory
 // and all of its subdirectories.
 func (prober *flexVolumeProber) initWatcher() error {
 	var err error
-	if prober.watcher, err = fsnotify.NewWatcher(); err != nil {
-		return fmt.Errorf("Error creating new watcher: %s", err)
+	if prober.watcher == nil {
```

Comment:
Do we need the nil check? Is initWatcher called more than once?

Reply:
The mock watcher is injected by creating an instance of flexVolumeProber with watcher already assigned. The nil check is to make sure the mock is not overridden.

Reply:
Ideally code shouldn't know whether or not it's working with a mock, doing so breaks the abstraction.
Can we always assign the watcher to the flexVolumeProber before we get here?

Reply:
We can initialize the watcher in GetDynamicPluginProber() (the prober constructor, essentially), but it would have to handle the possibility of an error somehow. Returning the error from a constructor seems strange, and we shouldn't handle the error here. IMO it's better to call it in Init().

Or, it gets passed into the constructor as a dependency. I don't think the caller should worry about the details of underlying watcher implementation, especially when most of the time it's the default watcher anyway.

From the perspective of this code, it's doing a safety check to make sure watcher is always initialized. There isn't any code that depends on the watcher being a mock or not.

Reply:
It's fine to return an error from the constructor, IMO.